Add TypeScript types

mbklein opened the pull request:
Adds a `/types` directory and the appropriate `tsconfig.json` and `package.json` entries to let compilers and language servers know where to find them.

Resolves #35

blimmer commented on types/processor.d.ts:
```diff
@@ -0,0 +1,30 @@
+export declare class Processor {
+  constructor(
+    url: string,
+    streamResolver: (id: string) => NodeJS.ReadableStream,
```
I don't think this is quite right based on the docs: https://github.com/samvera/node-iiif?tab=readme-ov-file#stream-resolver.

The first parameter will be something like `opts: { id: string, baseUrl: string }`.

The callback will need a bit more thought. We'll have to accept the parameters as described here: https://github.com/samvera/node-iiif/blob/main/README.md#stream-resolver

mbklein replied:
Ah, right, thanks for noticing that.

blimmer reviewed (commented):
This is looking good so far! I plugged this into an existing repo and saw that the `streamResolver` was throwing some type errors.

mbklein commented:
@blimmer I've updated the `StreamResolver` and `DimensionFunction` types with the correct signatures, and also added `Promise<>` wrappers to async functions, which I had forgotten before. I used `any` for the callback's return type, because the callback itself is internal to the package – all the `StreamResolver` needs to know is that it might receive a callback that expects to receive the `ReadableStream`.

Frankly, the `callback` is a vestige of the original implementation and I might remove it in the next major version.

blimmer commented:
We're getting closer, but the types still don't quite match up for me. Here's the file I'm testing against:

```ts
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { Processor } from "iiif-processor";
import type { IncomingMessage } from "node:http";

export type ImageResponse = { Body: IncomingMessage; ContentType?: string };

export async function handleImageRequest(): Promise<ImageResponse> {
  const s3 = new S3Client({});

  async function streamResolver(
    { id }: { id: string; baseUrl: string },
    callback: (stream: IncomingMessage) => Promise<unknown>,
  ) {
    const response = await s3.send(
      new GetObjectCommand({
        Bucket: "some-bucket-that-is-not-important",
        Key: id,
      }),
    );
    const stream = response.Body;
    if (!stream) {
      throw new Error(`Could not fetch object from S3: ${id}`);
    }

    try {
      return await callback(stream);
    } finally {
      stream.destroy();
    }
  }

  const userRequestUrl = "not-important";
  const url = decodeURIComponent(userRequestUrl);

  const processor = new Processor(url, streamResolver, {
    includeMetadata: true, // Keeping metadata to ensure orientation metadata is available (otherwise images can lose rotation)
    iiifVersion: 3,
    pathPrefix: "/",
  });

  const result = await processor.execute();
  const resp = {
    ContentType: result.contentType,
    Body: result.body,
  };

  return resp;
}
```

I redacted / stubbed some parameters in here so you can see a simple, self-contained example.

I agree RE: the callback piece, it feels a bit "old school" for newer Node patterns. I think it'd be great to show an example of fetching an image from S3 via the newer AWS SDK v3 since v2 will reach EOL in 2025.

mbklein commented:
I just pushed another commit with a couple more changes to the types, and an updated example in the README.  I didn't even realize it hadn't been updated to use v3 of the SDK.

I'm struggling with what to do about the callback as well, because I want to give the `StreamResolver` a chance to clean up after itself, which it can't do until after the Processor is done with the stream. The issue isn't that the StreamResolver has to be prepared for a callback not to be present, because it always will be. It's that the StreamResolver itself isn't _required_ to call the callback if it doesn't need to do any cleanup. So `callback?` is the wrong typedef – StreamResolver itself should really be an overloaded typedef for a function that accepts either one or two arguments.

The best answer is probably to get rid of the callback and replace it with an event (e.g. `processor.on('end', () => { cleanup... });`), but that's a breaking change. I might just bite the bullet and do it, though.

In the meantime, updating your StreamResolver to `return callback ? await callback(stream) : stream` will fix the existing typing issue.

mbklein commented:
I've added the overloaded typedef for `StreamResolver`. I still need to document it better, but `callback` is no longer a nullable argument so the implementation won't have to worry about it.

blimmer reviewed (approved):
Sorry for the delay - lgtm!
